kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/S3OutputStream.java

```diff
@@ -219,19 +219,26 @@ private MultipartUpload newMultipartUpload() throws IOException {
       initRequest.setSSECustomerKey(sseCustomerKey);
     }
 
+    return handleAmazonExceptions(
+      () -> new MultipartUpload(s3.initiateMultipartUpload(initRequest).getUploadId())
+    );
+  }
+
+  /**
+   * Return the given Supplier value, converting any thrown AmazonClientException object
+   * to an IOException object (containing the AmazonClientException object) and
+   * throw that instead.
+   * @param supplier The supplier to evaluate
+   * @param <T> The object type returned by the Supplier
+   * @return The value returned by the Supplier
+   * @throws IOException Any IOException or AmazonClientException thrown while
+   *                     retreiving the Supplier's value
+   */
+  private static <T> T handleAmazonExceptions(Supplier<T> supplier) throws IOException {
     try {
-      return new MultipartUpload(s3.initiateMultipartUpload(initRequest).getUploadId());
-    } catch (AmazonServiceException e) {
-      if (e.getErrorType() == ErrorType.Client) {
-        // S3 documentation states that this error type means there is a problem with the request
-        // and that retrying this request will not result in a successful response. This includes
-        // errors such as incorrect access keys, invalid parameter values, missing parameters, etc.
-        // Therefore, the connector should propagate this exception and fail.
-        throw new ConnectException("Unable to initiate MultipartUpload", e);
-      }
-      throw new IOException("Unable to initiate MultipartUpload.", e);
+      return supplier.get();
     } catch (AmazonClientException e) {
-      throw new IOException("Unable to initiate MultipartUpload.", e);
+      throw new IOException(e);
```

Comment:
should we be passing a message down here in order to not lose "Unable to initiate MultipartUpload." ?

Reply:
I removed the message intentionally because now this method is used in two places - during initiating multipart upload and during completion of multipart upload. So keeping the message would be misleading. I decided that having original exception passed as a cause to IOException will be enough to figure out what was the origin of the exception.

Do you think we still need to pass some message to IOException in addition to a "cause" exception?